### PR TITLE
discovery/kubernetes: optimize resolvePodRef

### PR DESCRIPTION
**Pull request by GiedriusS**
resolvePodRef is in a hot path:

```
ROUTINE ======================== github.com/prometheus/prometheus/discovery/kubernetes.(*Endpoints).resolvePodRef in discovery/kubernetes/endpoints.go
    2.50TB     2.66TB (flat, cum) 22.28% of Total
         .          .    447:func (e *Endpoints) resolvePodRef(ref *apiv1.ObjectReference) *apiv1.Pod {
         .          .    448:   if ref == nil || ref.Kind != "Pod" {
         .          .    449:           return nil
         .          .    450:   }
    2.50TB     2.50TB    451:   p := &apiv1.Pod{}
         .          .    452:   p.Namespace = ref.Namespace
         .          .    453:   p.Name = ref.Name
         .          .    454:
         .   156.31GB    455:   obj, exists, err := e.podStore.Get(p)
         .          .    456:   if err != nil {
         .          .    457:           level.Error(e.logger).Log("msg", "resolving pod ref failed", "err", err)
         .          .    458:           return nil
         .          .    459:   }
         .          .    460:   if !exists {
```

This is some low hanging fruit that we can easily optimize. The key of an object has format "namespace/name" so generate that inside of Prometheus itself and use pooling.

```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/discovery/kubernetes
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
                 │   olddisc    │               newdisc               │
                 │    sec/op    │   sec/op     vs base                │
ResolvePodRef-16   516.3n ± 17%   289.5n ± 7%  -43.92% (p=0.000 n=10)

                 │   olddisc    │              newdisc               │
                 │     B/op     │    B/op     vs base                │
ResolvePodRef-16   1168.00 ± 0%   24.00 ± 0%  -97.95% (p=0.000 n=10)

                 │  olddisc   │            newdisc             │
                 │ allocs/op  │ allocs/op   vs base            │
ResolvePodRef-16   2.000 ± 0%   2.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```

### CODE REVIEW

File: discovery/kubernetes/endpoints.go [hunk 1]
```diff
@@ -453,15 +455,43 @@ func (e *Endpoints) buildEndpoints(eps *apiv1.Endpoints) *targetgroup.Group {
 	return tg
 }
 
+var objKeyPool = sync.Pool{}
+
+func generateObjKey(namespace, name string) (string, *strings.Builder) {
+	var sb *strings.Builder
```
Comment (machine424) [inline discovery/kubernetes/endpoints.go:461]: I'm afraid a builder would be an overkill here and we can do without its overhead, we just want to return `namespace/name` (endpoints are namespaced)
You can use `namespacedName` from `kubernetes.go`

**Review by machine424 — COMMENTED** [review]
Good "discovery" (no wordplay intended ;))
`.Get()` should be calling `.GetByKey()`, but in our case, we don't need to carry the whole object around, as you mentioned we "implicitly" use `ns/name` as the key for namespaced resources.

We may want to do the same for `endpointsslice` or even check all `.Get()` and see if we can directly call `GetByKey`

**Comment by GiedriusS** [comment]
Thanks! I didn't see that function and it makes sense. I have updated other similar code paths to use namespacedName. Even better:

```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/discovery/kubernetes
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
                 │   olddisc    │              newdiscv2              │
                 │    sec/op    │   sec/op     vs base                │
ResolvePodRef-16   516.3n ± 17%   265.3n ± 4%  -48.62% (p=0.000 n=10)

                 │   olddisc    │             newdiscv2              │
                 │     B/op     │    B/op     vs base                │
ResolvePodRef-16   1168.00 ± 0%   16.00 ± 0%  -98.63% (p=0.000 n=10)

                 │  olddisc   │             newdiscv2              │
                 │ allocs/op  │ allocs/op   vs base                │
ResolvePodRef-16   2.000 ± 0%   1.000 ± 0%  -50.00% (p=0.000 n=10)
```

**Review by machine424 — APPROVED** [review]
lgtm, thanks for this.
(If we could have the same for `discovery/kubernetes/endpointslice.go`, it'd be great, otherwise I can take care of it)

**Review by machine424 — APPROVED** [review]
thanks again.

**Pull request merged**